crates/rustc_codegen_spirv/src/codegen_cx/constant.rs

```diff
@@ -374,23 +397,65 @@ impl<'tcx> CodegenCx<'tcx> {
-                self.constant_bool(DUMMY_SP, self.read_alloc_val(alloc, offset, 1) != 0)
-            }
-            SpirvType::Integer(width, _) => {
-                let v = self.read_alloc_val(alloc, offset, (width / 8) as usize);
-                self.constant_int(ty, v as u64)
-            }
-            SpirvType::Float(width) => {
-                let v = self.read_alloc_val(alloc, offset, (width / 8) as usize);
-                match width {
-                    32 => self.constant_f32(DUMMY_SP, f32::from_bits(v as u32)),
-                    64 => self.constant_f64(DUMMY_SP, f64::from_bits(v as u64)),
-                    other => self
+            SpirvType::Bool
+            | SpirvType::Integer(..)
+            | SpirvType::Float(_)
+            | SpirvType::Pointer { .. } => {
+                let size = ty_concrete.sizeof(self).unwrap();
+                let primitive = match ty_concrete {
+                    SpirvType::Bool => Primitive::Int(Integer::fit_unsigned(0), false),
+                    SpirvType::Integer(int_size, int_signedness) => {
+                        let integer = match int_size {
+                            8 => Integer::I8,
+                            16 => Integer::I16,
+                            32 => Integer::I32,
+                            64 => Integer::I64,
+                            128 => Integer::I128,
+                            other => {
+                                self.tcx
+                                    .sess
+                                    .fatal(&format!("invalid size for integer: {}", other));
+                            }
+                        };
+                        Primitive::Int(integer, int_signedness)
+                    }
+                    SpirvType::Float(float_size) => match float_size {
+                        32 => Primitive::F32,
+                        64 => Primitive::F64,
+                        other => {
+                            self.tcx
+                                .sess
+                                .fatal(&format!("invalid size for float: {}", other));
+                        }
+                    },
+                    SpirvType::Pointer { .. } => Primitive::Pointer,
+                    unsupported_spirv_type => bug!(
+                        "invalid spirv type internal to create_alloc_const2: {:?}",
+                        unsupported_spirv_type
+                    ),
+                };
+                // alloc_id is not needed by read_scalar, so we just use 0. If the context
+                // refers to a pointer, read_scalar will find the the actual alloc_id. It
+                // only uses the input alloc_id in the case that the scalar is uninitilized
+                // as part of the error output
+                // tldr, the pointer here is only needed for the offset
+                let scalar = match alloc
+                    .read_scalar(self, Pointer::new(AllocId(0), *offset), size)
```

Comment:
If you pass `!0` instead of `0` in the `AllocId(...)`, it might be harder for this to go wrong (cc @oli-obk).

Reply:
we use `0` in the compiler, too, so it seems fine. If we change anything here it'll be changing the datastructure in a way that needs code changes, so there won't be silent breakage or something.

Reply:
In lieu of the above, I just left it as 0